tfctl/show_plan.go

```diff
@@ -0,0 +1,82 @@
+package tfctl
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/hashicorp/terraform-exec/tfexec"
+	infrav1 "github.com/weaveworks/tf-controller/api/v1alpha1"
+	"github.com/weaveworks/tf-controller/utils"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// ShowPlan displays the plan for the given Terraform resource
+func (c *CLI) ShowPlan(out io.Writer, resource string) error {
+	key := types.NamespacedName{
+		Name:      resource,
+		Namespace: c.namespace,
+	}
+	terraform := &infrav1.Terraform{}
+	if err := c.client.Get(context.TODO(), key, terraform); err != nil {
+		return fmt.Errorf("resource %s not found", resource)
+	}
+
+	if terraform.Status.Plan.Pending == "" {
+		fmt.Fprintln(out, "There is no plan pending.")
+		return nil
+	}
+
+	planKey := types.NamespacedName{
+		Name:      fmt.Sprintf("tfplan-default-%s", resource),
+		Namespace: c.namespace,
+	}
+
+	planSecret := &corev1.Secret{}
+	if err := c.client.Get(context.TODO(), planKey, planSecret); err != nil {
+		return fmt.Errorf("plan for resource %s not found", resource)
+	}
+
+	data, err := utils.GzipDecode(planSecret.Data["tfplan"])
+	if err != nil {
+		return fmt.Errorf("failed to decode plan for resources %s: %s", resource, err)
+	}
+
+	tmpDir, err := ioutil.TempDir(os.TempDir(), "tfctl")
+	if err != nil {
+		return err
+	}
+
+	planFile, err := ioutil.TempFile(tmpDir, "tfctl-plan-")
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := os.RemoveAll(tmpDir); err != nil {
+			fmt.Fprintf(out, "failed to remove temporary directory %s: %s", tmpDir, err)
+		}
+	}()
+
+	if err := os.WriteFile(planFile.Name(), data, 0644); err != nil {
```

Comment:
This code is a bit odd, because you create a file with `ioutil.TempFile` which creates with `0600`, then write to the file with a different set of perms `0644`.

In the middle you should probably close `planFile`?

If you don't need `0644` then you can just write to the file with `os.Write(data)`?

Alternatively, maybe you don't need the temporary _filename_ and you could just [filepath.Join(tmpdir, "tfctl-plan-name")](https://pkg.go.dev/path/filepath#Join).

As you're creating a new tempdir every time, it feels unlikely that you'd get a collision?